### PR TITLE
context keys: implement a new parser (and a scanner/lexer) for 'when' clauses

### LINKED ISSUE #91473: Add support for parenthesis in "when" conditions
- VSCode Version: 1.42.0
- OS Version: Ubuntu 18.04.3

The `when` expression that contains parentheses which we set for a keybinding defined in the contribution point of our extension results into a corrupted (invalidly transformed expression in user's keybinding settings), e.g. the following declaration:
```json
{
    "command": "extension.helloWorld",
    "key": "enter",
    "when": "editorLangId == rust && !suggestWidgetVisible && (!vim.active || vim.mode == 'Insert')"
}
```
produces the following invalid when expression:

Steps to Reproduce:

1. Go to the [following repo with the reproduction](https://github.com/Veetaha/report-vscode-corrupted-when-codition-with-parens)
2. Follow the steps in README.md

Does this issue occur when all extensions are disabled?: Yes

### PR DESCRIPTION
**Pull request by ulugbekna**
This PR introduces a new parser for "when" [clause](https://code.visualstudio.com/api/references/when-clause-contexts) contexts. 

The new parser has a (semi-)formal syntax grammar provided at the end of this description. As a result it has more strict parsing rules that we did before. 

When I say "key" or "key name" below, I mean the identifier that is looked up from the when-clause's context.

## New features introduced

- support for parenthesis (fixes https://github.com/microsoft/vscode/issues/91473) 

   Data (based on ~16040 when-clauses from ~18100 most popular extensions): at least 35 when-clauses in the wild already use parentheses already (ie people generally just assume parentheses support)

- support for all regex flags, the old parser supported only `i`; the new one supports all: i g s m y u
   
   Data: there're at least 21 when-clauses in the wild that use regex flags besides `i`. 

- syntax grammar that users can address to write when-clauses correctly

For example, if a user wrote a when-clause: 

```
!isEditor == 'true'
``` 

the old parser would recognize `!isEditor` as a key rather than a negation because there was no clear specification of how a when-clause was allowed to look like. The new parser shows a parsing error because we support only negation of a key (we can change this behavior, but it is still better than old parser's behavior). 

- lexical and syntax error reporting

### Regexes 

- we stop supporting regex that do not escape slashes, e.g., writing `foo =~ /file:////` was supported in the old parser, while the new parser requires `foo =~ /file:\\/\\/\\/`. This looks uglier but it's very hard to lex a regex with unescaped slashes, e.g., javascript also can't parse `const foo = /foo//`. 

   Data: around 50 when-clauses that use unescaped slashes in their regexes. 

Note: the old parser had its own problems parsing such expressions, where `foo =~ /&&/` would be parsed as `(foo =~ /invalid/) && /`, i.e., an AND expression with `/` as a key. 

### Comparison operators

The old parser allowed comparison operators `>` and `<` in key names, e.g., `vim.use<C-r>`, but only if it's not within a comparison -- `vim.usesOf<C-r> < 1` would all be parsed as a single key name, ie `{ "key" : "vim.usesOf<C-r> < 1"}`. 

The new parser allows `<` and `>` in key names to not break existing key names such as `vim.use<C-r>`, but then we don't allow non-space-separated comparison expressions such as `foo<1`. We still support `foo<=1` as normal comparison (though at some time cost). 

- [ ] From 16K when-clauses I've observed, not a single use non-space separated comparison currently. I propose that we disallow comparison without spaces around. 

### Space-separated strings

The old parser allowed `foo == Bar Jar` as `foo == 'Bar Jar'`, 
the new parser reports an error in such a case. One can wrap the space-separated string in single-quotes. 

Data: I found ~9 when-clauses with space-separated but unquoted strings. 

## Syntax grammar

(Uses [EBNF](https://en.wikipedia.org/wiki/Extended_Backus%E2%80%93Naur_form))

```ebnf

expression ::= or

or ::= and { '||' and }*

and ::= term { '&&' term }*

term ::=
	| '!' (KEY | 'true' | 'false')
	| primary

primary ::=
	| 'true'
	| 'false'
	| '(' expression ')'
	| KEY '=~' REGEX
	| KEY [ ('==' | '!=' | '<' | '<=' | '>' | '>=' | 'not' 'in' | 'in') value ]

value ::=
	| 'true'
	| 'false'
	| 'in'     
	| KEY
	| SINGLE_QUOTED_STR
	| EMPTY_STR
```

### CODE REVIEW

File: src/vs/platform/contextkey/common/contextkey.ts [hunk 1]
```diff
@@ -129,6 +131,32 @@ export abstract class ContextKeyExpr {
...
+				warning.push('Lexing errors:\n\n');
+				parser.lexingErrors.forEach(token => warning.push(Scanner.reportError(token), '\n'));
+			}
+
+			if (parser.parsingErrors.length > 0) {
+				if (parser.lexingErrors.length > 0) { warning.push('\n\n'); } // separate lexing errors from parsing errors
+				warning.push('Parsing errors:\n\n');
+				parser.parsingErrors.forEach(message => warning.push(`${message}`, '\n'));
+			}
+			console.warn(warning.join(''));
```
Comment (alexdima) [inline src/vs/platform/contextkey/common/contextkey.ts:149]: We shouldn't do console.warn here, this is too low level. I suggest we create a new variant of `ContextKeyExpr.deserialize`, perhaps `ContextKeyExpr.deserializeWithErrors` which would return the errors and then adopt that in the various callers. Those callers might have a chance to report more useful information, like for example the extension id that has the problematic context key, or the keybinding item, etc.

File: src/vs/platform/contextkey/common/contextkey.ts [hunk 2]
```diff
@@ -241,6 +269,370 @@ export abstract class ContextKeyExpr {
...
+				return '||';
+			case TokenType.EOF:
+				return 'EOF';
+			default:
+				throw illegalState(`all other tokens must have a lexeme : ${JSON.stringify(token)}`);
+		}
+	}
+
+	private _check(type: TokenType) {
+		return !this._isAtEnd() && this._peek().type === type;
```
Comment (alexdima) [inline src/vs/platform/contextkey/common/contextkey.ts:624]: The check for `!this._isAtEnd()` doesn't seem necessary here

File: src/vs/platform/contextkey/common/contextkey.ts [hunk 3]
```diff
@@ -241,6 +269,370 @@ export abstract class ContextKeyExpr {
...
+			return ''; // this allows "when": "foo == " which's used by existing extensions
+		}
+	}
+
+	// careful: this can throw if current token is the initial one (ie index = 0)
+	private _previous() {
+		return this._tokens[this._current - 1];
+	}
+
+	private _match(...tokens: TokenType[]) {
```
Comment (alexdima) [inline src/vs/platform/contextkey/common/contextkey.ts:540]: AFAIK functions with for loops rarely get inlined. I'd like to suggest to create two versions, `_matchOne` and `_matchAny` and use each on of them as needed directly. I think this is a low hanging perf improvement.

File: src/vs/platform/contextkey/common/scanner.ts [hunk 4]
```diff
@@ -0,0 +1,387 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { CharCode } from 'vs/base/common/charCode';
+import { illegalArgument } from 'vs/base/common/errors';
+
+export enum TokenType {
```
Comment (alexdima) [inline src/vs/platform/contextkey/common/scanner.ts:9]: I would suggest to convert this to a `const enum` and let TS auto-number it for you. Switching to a const enum will lead to TS inlining the numbers in the generated JS code and in the comparisons being a lot faster at runtime. This is also a low hanging perf improvement.

File: src/vs/platform/contextkey/common/scanner.ts [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+	Str = 'Str',
+
+	QuotedStr = 'QuotedStr',
+
+	Error = 'ErrorToken',
+
+	EOF = 'EOF'
+}
+
+export type Token = {
```
Comment (alexdima) [inline src/vs/platform/contextkey/common/scanner.ts:45]: Once you switch to a const enum, you can write this as an or type (e.g. `type Token = { type: TokenType.Str; offset: number; lexeme: string} | ... ) and add strong typing around which tokens have a `lexeme` or not. This would allow removing a lot of `!` from the code.

File: src/vs/platform/contextkey/common/contextkey.ts [hunk 6]
```diff
@@ -241,6 +269,370 @@ export abstract class ContextKeyExpr {
...
+		return this._error(`Unexpected '${this._nameToken(token)}' at offset ${token.offset}.`);
+	}
+
+	// TODO@ulugbekna: the whole error reporting needs reworking - especially, when we introduce it to package.json linting
+	private _error(errMsg: string) {
+		this._parsingErrors.push(errMsg);
+		return new ParseError();
+	}
+
+	private _nameToken(token: Token): string {
```
Comment (alexdima) [inline src/vs/platform/contextkey/common/contextkey.ts:580]: This could be extracted to a function in the same file where `Token` is defined (because it doesn't seem to use any properties).

File: src/vs/platform/contextkey/common/scanner.ts [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+		return this._token(TokenType.RegexStr, true);
+	}
+
+	// invariant: this must not fail if at end of `this._value`
+	private _eatWhitespace() {
+		Scanner._whitespace.lastIndex = this._current;
+		const match = Scanner._whitespace.exec(this._input);
+		if (match) {
+			this._current += match[0].length;
+		}
```
Comment (alexdima) [inline src/vs/platform/contextkey/common/scanner.ts:371]: I believe this method doesn't do anything, because `/^\s*` matches any input and always returns a first match of length 0. I tried commenting out this method and it looks that all tests still pass. I think we should add a test that asserts that parsing `" a"` (leading whitespace) works correctly.

File: src/vs/platform/contextkey/common/scanner.ts [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+		if (this._input[this._current] !== expected) {
+			return false;
+		}
+		this._current++;
+		return true;
+	}
+
+	private _advance(): string {
+		return this._input[this._current++];
```
Comment (alexdima) [inline src/vs/platform/contextkey/common/scanner.ts:204]: It would be better to use `String.charCodeAt` instead of the array access and to work everywhere with the `CharCode` const enum. That would be also a low hanging perf improvement.

**Review by alexdima — COMMENTED** [review]
Nice! I've left some comments.

File: src/vs/platform/contextkey/common/scanner.ts (hunk 7 above)
Comment (ulugbekna) [reply src/vs/platform/contextkey/common/scanner.ts]: You're right, `_eatWhitespace` was broken completely. Things worked because `next()` fn also handled whitespace.

**Comment by ulugbekna** [comment]
@alexdima Thanks for the review. I will re-request a review as soon as I address all your comments.

Meanwhile, I (1) rewrote the scanner in eager style and (2) rewrote lexing of a key name using a regex. The latter speeded up things significantly - where we saw ~15-17ms, now I see ~10ms. (JS Profiler tool is great)

File: src/vs/platform/contextkey/common/contextkey.ts (hunk 2 above)
Comment (ulugbekna) [reply src/vs/platform/contextkey/common/contextkey.ts:624]: Great catch!

File: src/vs/platform/contextkey/common/scanner.ts (hunk 4 above)
Comment (ulugbekna) [reply src/vs/platform/contextkey/common/scanner.ts:9]: interestingly, letting TS to auto-generate values for enum cases for you as below blows up running time from ~10ms to ~200ms

```ts
export const enum TokenType {
	LParen,
	RParen,
...
} 
```

image

I will experiment with other ways to speed this up.

**Comment by ulugbekna** [comment]
@alexdima This PR leaves the old parser as the default one, so I propose that we merge as long as 
1) you're okay with it
2) tests pass 

All resolved comments are addressed. The unresolved comments: 

> I think we should store the errMsg as the .message property of the error, otherwise all these ParseError instances are indistinguishable from each other

During my current linting work, I may rework how error reporting is done in general -- the main point is that ideally we wouldn't report strings as errors, but objects that are then converted to strings with appropriate messages. 

> It would be better to use String.charCodeAt instead of the array access and to work everywhere with the CharCode const enum from src/vs/base/common/charCode.ts. That would be also a low hanging perf improvement.

I will address this in a later PR that aims at perf improvements in general, if that's okay.

**Review by alexdima — APPROVED** [review]
:+1: Nice!
